platform/glfw/test_writer.cpp

```diff
@@ -0,0 +1,187 @@
+#include "test_writer.hpp"
+
+#include <rapidjson/prettywriter.h>
+#include <rapidjson/stringbuffer.h>
+#include <rapidjson/writer.h>
+#include <ghc/filesystem.hpp>
+
+#include <fstream>
+
+using Writer = rapidjson::PrettyWriter<rapidjson::StringBuffer>;
+
+class TestOperation {
+public:
+    virtual ~TestOperation() = default;
+
+    virtual void serialize(Writer& writer) const = 0;
+};
+
+class SetCamera final : public TestOperation {
+public:
+    SetCamera(const mbgl::CameraOptions& camera_) : camera(camera_) {}
+
+    void serialize(Writer& writer) const override {
+        if (camera.zoom) {
+            writer.StartArray();
+            writer.String("setZoom");
+            writer.Double(*camera.zoom);
+            writer.EndArray();
+        }
+
+        if (camera.bearing && *camera.bearing != 0.) {
+            writer.StartArray();
+            writer.String("setBearing");
+            writer.Double(*camera.bearing);
+            writer.EndArray();
+        }
+
+        if (camera.pitch && *camera.pitch != 0.) {
+            writer.StartArray();
+            writer.String("setPitch");
+            writer.Double(*camera.pitch);
+            writer.EndArray();
+        }
+
+        writer.StartArray();
+        writer.String("setCenter");
+        writer.StartArray();
+        writer.Double(camera.center->longitude());
+        writer.Double(camera.center->latitude());
+        writer.EndArray();
+        writer.EndArray();
+    }
+
+private:
+    mbgl::CameraOptions camera;
+};
+
+class SetStyle final : public TestOperation {
+public:
+    SetStyle(const mbgl::style::Style& style) : url(style.getURL()) {}
+
+    void serialize(Writer& writer) const override {
+        writer.StartArray();
+        writer.String("setStyle");
+        writer.String(url);
+        writer.EndArray();
+    }
+
+private:
+    std::string url;
+};
+
+class SetInitialSize final : public TestOperation {
+public:
+    SetInitialSize(const mbgl::Size& size) : width(size.width), height(size.height) {}
+
+    void serialize(Writer& writer) const override {
+        writer.Key("width");
+        writer.Int(width);
+
+        writer.Key("height");
+        writer.Int(height);
+    }
+
+private:
+    uint32_t width;
+    uint32_t height;
```

Comment:
those too class members are left uninitialized after construction

Reply:
Initialized here, `SetInitialSize(const mbgl::Size& size) : width(size.width), height(size.height) {}`, class has not default constructor.